File: android/src/main/kotlin/net/spross/app/ui/BoxRowMenu.kt
package net.spross.app.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlin.math.max
import kotlin.math.roundToInt
import net.spross.app.AppModel
import net.spross.app.CardDisplay
import net.spross.app.Chrome
import net.spross.app.forgetCard
import net.spross.app.ownWords
import net.spross.app.removeOwnWord
import net.spross.kern.box.BoxBrowser
import net.spross.kern.box.BoxEngine
import net.spross.kern.box.CardRowState
import net.spross.kern.model.Card

/**
 * Everything a learner might want to do to ONE word, on the long press every box row carries.
 *
 * Only what applies, and in one order wherever a row is drawn: where the word stands in the
 * queue, then where it stands in the schedule, then what can be made of it, then what is
 * wrong with it, and last — set apart — taking it out for good.
 *
 * WHICH of those apply is kern's answer, not this file's: [BoxBrowser.cardRowState] rules on
 * the standing, and it is asked with `packOffered = true` because the MENU always offers the
 * queue. The row's own right edge is the surface that varies (a shelf packs by the shelf);
 * a menu opened by name is the learner naming this one word.
 *
 * The session card's menu is deliberately NOT this one ([ReportableCard]): a round is no
 * place to reorganize the box.
 */
@Composable
internal fun BoxRowMenu(
    model: AppModel,
    card: Card,
    expanded: Boolean,
    onDismiss: () -> Unit,
    /** Opens the own-word form on the draft handed to it — a copy of this card, or an edit. */
    onWriteOwn: ((OwnWordDraft) -> Unit)?,
) {
    val chrome = model.chrome
    val state = model.box ?: return
    val standing = BoxBrowser.cardRowState(state, card.id, packOffered = true)
    val scheduled = standing is CardRowState.Sleeping || standing is CardRowState.Standing
    val own = model.ownWords.firstOrNull { it.id == card.id }
    val stamp = state.joinStamp

    CardMenu(
        model = model,
        card = card,
        expanded = expanded,
        // Nothing was being answered here, so the report carries no typed answer.
        learnerInput = "",
        onDismiss = onDismiss,
        before = { close ->
            // why: the row has no width for any of this and the box has no card to
            // reveal, so the long press is where a word explains itself — the same
            // lines the session's reveal wears, above every action rather than among them.
            WordCardLines(model, card, chrome)
            when (standing) {
                CardRowState.PackOffered -> MenuAction(chrome.boxCardPack) {
                    close()
                    model.updateBox { BoxEngine.enqueue(it, listOf(card.id)) }
                }

                is CardRowState.Packed -> MenuAction(chrome.boxCardUnpack) {
                    close()
                    model.updateBox { BoxEngine.dequeue(it, card.id) }
                }

                is CardRowState.Standing -> MenuAction(chrome.boxCardSleep) {
                    close()
                    model.updateBox { BoxEngine.setSuspended(it, card.id, true, model.now()) }
                }

                CardRowState.Sleeping -> MenuAction(chrome.boxCardWake) {
                    close()
                    model.updateBox { BoxEngine.setSuspended(it, card.id, false, model.now()) }
                }

                CardRowState.Plain -> Unit
            }
            if (scheduled) {
                MenuAction(chrome.boxCardForget) { close(); model.forgetCard(card.id) }
            }
            onWriteOwn?.let { write ->
                // Both sides carried over: the learner wants THIS pair in their own words,
                // not a blank sheet with the search box's guess in it.
                MenuAction(chrome.boxCardOwnFrom) {
                    close()
                    write(
                        OwnWordDraft(
                            known = card.source.text,
                            learning = card.target.text,
                            emoji = card.emoji.orEmpty(),
                        ),
                    )
                }
                own?.let { word ->
                    MenuAction(chrome.boxOwnWordEdit) {
                        close()
                        write(OwnWordDraft.of(word, stamp.source, stamp.target))
                    }
                }
            }
        },
        after = { close ->
            // why: only a word the learner wrote is theirs to delete — a catalog word can be
            // put to sleep, never removed, so it grows no such entry at all. Last and in the
            // error color: the one irreversible thing in the menu.
            own?.let { word ->
                MenuAction(chrome.boxOwnWordRemove, destructive = true) {
                    close()
                    model.removeOwnWord(word.id)
                }
            }
        },
    )
}

/**
 * What the long press hands over beyond the row itself, in the reveal's own order:
 * the other words for the same thing ("auch: …", exactly as [CardDisplay.alsoLine]
 * offers them mid-round), the catalog's gloss in the line the reveal wears
 * ([PauseLine]), and how long the word keeps.
 *
 * Draws nothing at all — not even its divider — where the word carries none of the
 * three, so the menu is not opened by an empty header.
 */
@Composable
private fun WordCardLines(model: AppModel, card: Card, chrome: Chrome) {
    val also = CardDisplay.alsoLine(card.target, chrome, card.target.text)
    val note = card.target.note ?: card.source.note
    // why: a card with no schedule reports 0 stability, and "Hält 0 Tage" would be a
    // fact about the engine rather than about the word — the row itself says nothing
    // about an untouched word either, so neither does this.
    val lasts = model.cardGrowth(card.id)?.stability
        ?.takeIf { it > 0.0 }?.let { max(1, it.roundToInt()) }
    if (also == null && note == null && lasts == null) return
    Column(
        Modifier.widthIn(max = 280.dp).padding(Theme.spacing.md),
        verticalArrangement = Arrangement.spacedBy(Theme.spacing.xs),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        also?.let { CardLine(it, Modifier.fillMaxWidth()) }
        note?.let { PauseLine(it) }
        lasts?.let {
            val form = if (it == 1) chrome.boxCardLastsOne else chrome.boxCardLasts
            CardCue(form.format(it), Modifier.fillMaxWidth())
        }
    }
    HorizontalDivider()
}

/** One entry of a word's menu; [destructive] is the only one that is not undoable. */
@Composable
internal fun MenuAction(label: String, destructive: Boolean = false, onClick: () -> Unit) {
    DropdownMenuItem(
        text = {
            Text(
                label,
                color = if (destructive) {
                    MaterialTheme.colorScheme.error
                } else {
                    MaterialTheme.colorScheme.onSurface
                },
            )
        },
        onClick = onClick,
    )
}

File: App/Sources/Screens/BoxRowMenu.swift
import SwiftUI
import SprossKern

/// Which sheet a row's menu asks the row to raise. The menu itself lives inside a
/// `contextMenu` builder, which is no part of the view hierarchy a sheet can be
/// presented from — so the choice travels back out and the row presents it.
enum BoxRowSheet: String, Identifiable {
    case report, editReport, ownFrom, editOwnWord

    var id: String { rawValue }
}

/// Everything a learner might want to do to one word, as a long press on its Box row.
///
/// Only what applies, and in one fixed order: where it stands in the box first
/// (packing, sleep, forgetting), then what can be MADE of it, then what is wrong
/// with it, and deleting last because it is the one entry that cannot be taken back.
/// WHERE the word stands is the box's own ruling (`BoxBrowser.cardRowState`), asked
/// with packing offered — this menu can always pack a single word, whatever the row
/// behind it draws.
///
/// The session card's menu is deliberately NOT this one (`SessionView.cardMenu`):
/// a round is no place to reorganize the box.
struct BoxRowMenu: View {
    let model: AppModel
    let card: Card
    /// Raise a sheet the menu cannot present itself.
    let open: (BoxRowSheet) -> Void

    var body: some View {
        standing
        Button("box.card.ownFrom", systemImage: "doc.on.doc") { open(.ownFrom) }
        if model.isOwnWord(card.id) {
            Button("box.own.word.edit", systemImage: "pencil") { open(.editOwnWord) }
        }
        reporting
        if model.isOwnWord(card.id) {
            Button("box.own.word.remove", systemImage: "trash", role: .destructive) {
                model.removeOwnWord(card.id)
            }
        }
    }

    /// Packing, sleep and forgetting — the three things that move a word's standing
    /// rather than its content. A card the join does not hold offers none of them.
    @ViewBuilder
    private var standing: some View {
        let state = model.cardRowState(card.id, packOffered: true)
        switch onEnum(of: state) {
        case .packOffered:
            Button("box.card.pack", systemImage: "tray.and.arrow.down") {
                model.enqueueCard(card.id)
            }
        case .packed:
            Button("box.card.unpack", systemImage: "tray.and.arrow.up") {
                model.dequeue(cardID: card.id)
            }
        case .standing:
            Button("box.card.sleep", systemImage: "moon.zzz") {
                model.setSuspended(cardID: card.id, suspended: true)
            }
        case .sleeping:
            Button("box.card.wake", systemImage: "sun.max") {
                model.setSuspended(cardID: card.id, suspended: false)
            }
        case .plain:
            EmptyView()
        }
        if isScheduled(state) {
            Button("box.card.forget", systemImage: "arrow.counterclockwise") {
                model.forget(cardID: card.id)
            }
        }
    }

    /// Whether the box has ever asked this word — the one state where there is
    /// progress to drop.
    private func isScheduled(_ state: CardRowState) -> Bool {
        switch onEnum(of: state) {
        case .standing, .sleeping: return true
        case .packOffered, .packed, .plain: return false
        }
    }

    @ViewBuilder
    private var reporting: some View {
        if model.reportedIssue(for: card.id) == nil {
            Button("report.action", systemImage: "exclamationmark.bubble") { open(.report) }
        } else {
            Button("report.edit", systemImage: "text.bubble") { open(.editReport) }
            Button("report.dismiss", systemImage: "checkmark.bubble") {
                model.dismissReportedIssue(cardID: card.id)
            }
        }
    }
}
